File: python/network/nesting.py
import torch


def flatten(nested_input, traverse_function):
    flattened_input = []
    groups = []
    operators = {}
    for value, group_index, operator in traverse_function(nested_input):
        flattened_input.append(value)
        groups.append(group_index)
        operators[group_index] = operator
    return flattened_input, groups, operators


def traverse_filter(nested_input, path_list=None):
    if path_list is None:
        path_list = []

    for i, node in enumerate(nested_input):
        path_list.append(str(i))
        if node["IsLeaf"]:
            yield (
                [
                    node["Condition"]["Field"],
                    node["Condition"]["Operation"],
                    node["Condition"]["Value"],
                ],
                tuple(path_list),
                0,
            )
        else:
            yield from traverse_filter_node(
                node["Operands"], path_list, node["LogicalOperator"]
            )
        path_list.pop()  # Backtrack


def traverse_filter_node(nested_input, path_list, operator):
    for i, node in enumerate(nested_input):
        if node["IsLeaf"]:
            yield (
                [
                    node["Condition"]["Field"],
                    node["Condition"]["Operation"],
                    node["Condition"]["Value"],
                ],
                tuple(path_list),
                operator,
            )
        else:
            path_list.append(str(i))
            yield from traverse_filter_node(
                node["Operands"], path_list, node["LogicalOperator"]
            )
            path_list.pop()  # Backtrack


def traverse_filter_v2(nested_input, path_list=None):
    if path_list is None:
        path_list = []

    for i, node in enumerate(nested_input):
        path_list.append(i)
        if node["IsLeaf"]:
            yield (
                [
                    node["Condition"]["Field"],
                    node["Condition"]["Operation"],
                    node["Condition"]["Value"],
                ],
                tuple(path_list),
                0,
            )
        else:
            yield from traverse_filter_node_v2(
                node["Operands"], path_list, node["LogicalOperator"]
            )
        path_list.pop()  # Backtrack


def traverse_filter_node_v2(nested_input, path_list, operator):
    for i, node in enumerate(nested_input):
        if node["IsLeaf"]:
            yield (
                [
                    node["Condition"]["Field"],
                    node["Condition"]["Operation"],
                    node["Condition"]["Value"],
                ],
                tuple(path_list),
                operator,
            )
        else:
            path_list.append(i)
            yield from traverse_filter_node_v2(
                node["Operands"], path_list, node["LogicalOperator"]
            )
            path_list.pop()  # Backtrack


def is_prefix(prefix, test):
    return len(prefix) <= len(test) and test[: len(prefix)] == prefix


def add_to_stack(
    element,
    current_combination,
    current_groups,
    group_index,
    operators: dict,
    combine_function: callable,
):
    if current_groups and group_index == current_groups[-1]:
        current_combination[-1].append(element)
        return

    # Break down stack until we find the right place
    break_down_stack(
        current_combination, current_groups, group_index, operators, combine_function
    )

    # After breaking down, we know where to add
    if current_groups and group_index == current_groups[-1]:
        current_combination[-1].append(element)
    else:
        current_combination.append([element])
        current_groups.append(group_index)


def break_down_stack(
    current_combination,
    current_groups,
    group_index,
    operators: dict,
    combine_function: callable,
):
    while current_groups and not is_prefix(current_groups[-1], group_index):
        current_group = current_groups.pop()
        combined = combine_function(
            current_combination.pop(), operators.get(current_group)
        )
        reduced_index = current_group[:-1]
        add_to_stack(
            combined,
            current_combination,
            current_groups,
            reduced_index,
            operators,
            combine_function,
        )


def reduce(
    flattened_input: torch.Tensor,
    groups: list[tuple[str]],
    operators: dict,
    combine_function: callable,
) -> torch.Tensor:

    current_combination = []
    current_groups = []
    current_batch_index = 0

    for i in range(len(groups)):
        if not current_groups:
            current_groups.append((str(current_batch_index),))
            current_batch_index += 1
            current_combination.append([])

        group_index = groups[i]
        add_to_stack(
            flattened_input[i],
            current_combination,
            current_groups,
            group_index,
            operators,
            combine_function,
        )

    break_down_stack(
        current_combination, current_groups, (), operators, combine_function
    )
    if not current_combination:
        return []
    return current_combination.pop()


def reduce_v2(
    flattened_input: torch.Tensor,
    groups: list[tuple[int]],
    operators: dict,
    combine_function: callable,
) -> torch.Tensor:

    current_combination = []
    current_groups = []
    current_batch_index = 0

    for i in range(len(groups)):
        if not current_groups:
            current_groups.append((current_batch_index,))
            current_batch_index += 1
            current_combination.append([])

        group_index = groups[i]
        add_to_stack(
            flattened_input[i],
            current_combination,
            current_groups,
            group_index,
            operators,
            combine_function,
        )

    break_down_stack(
        current_combination, current_groups, (), operators, combine_function
    )
    if not current_combination:
        return []
    return current_combination.pop()


def vectorize_amount_data(
    amount_data: dict, device: torch.device = None, dtype: torch.dtype = None
) -> torch.Tensor:
    factory_kwargs = {"device": device, "dtype": dtype}
    return torch.tensor(
        [
            amount_data["Amount"]["Min"],
            amount_data["Amount"]["Max"],
            amount_data["FromPosition"],
        ],
        **factory_kwargs,
    )


def vectorize_attack_data(
    attack_data: dict, device: torch.device = None, dtype: torch.dtype = None
) -> torch.Tensor:
    factory_kwargs = {"device": device, "dtype": dtype}
    return torch.tensor(
        [attack_data["AttackTarget"], attack_data["Damage"]], **factory_kwargs
    )


def vectorize_discard_data(
    discard_data: dict, device: torch.device = None, dtype: torch.dtype = None
) -> torch.Tensor:
    factory_kwargs = {"device": device, "dtype": dtype}
    return torch.tensor([discard_data["TargetSource"]], **factory_kwargs)


def vectorize_return_to_deck_type_data(
    return_to_deck_type_data: dict,
    device: torch.device = None,
    dtype: torch.dtype = None,
) -> torch.Tensor:
    factory_kwargs = {"device": device, "dtype": dtype}
    return torch.tensor(
        [
            return_to_deck_type_data["ReturnToDeckType"],
            return_to_deck_type_data["FromPosition"],
        ],
        **factory_kwargs,
    )


def vectorize_player_target_data(
    player_target_data: dict, device: torch.device = None, dtype: torch.dtype = None
) -> torch.Tensor:
    factory_kwargs = {"device": device, "dtype": dtype}
    return torch.tensor([player_target_data["PlayerTarget"]], **factory_kwargs)


def flatten_instructions(
    instructions: list[list[dict]],
    device: torch.device = None,
    dtype: torch.dtype = None,
) -> tuple[
    torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor, list[dict], torch.Tensor
]:
    """
    Flatten the instructions
    Args:
        instructions: batch of instructions lists (batch_size, num_instructions*)
    Returns:
        tuple of
            tensor of instruction types dim (total number of instructions)
            tensor of instruction indices (batch_index, instruction_index) dim  (total number of instructions, 2)
            tensor of instruction data type except filter data dim (total number of instruction data)
            tensor of instruction data type indices (batch_index, instruction_index, data_index) dim (total number of instruction data, 3)
            tuple of instruction data tensor lists. One tuple entry for each instruction data type.
            tuple of tensor instruction data indices lists.
    """
    factory_kwargs = {"device": device, "dtype": dtype}
    instruction_types = []
    instruction_indices = []
    instruction_data_types = []
    instruction_data_type_indices = []
    instruction_data = ([], [], [], [], [], [])
    instruction_data_indices = ([], [], [], [], [], [])
    for batch_index, batch_instructions in enumerate(instructions):
        for instruction_index, instruction in enumerate(batch_instructions):
            instruction_types.append(instruction["InstructionType"])
            instruction_indices.append((batch_index, instruction_index))
            for data_index, data in enumerate(instruction["Data"]):
                data_type = data["InstructionDataType"]
                instruction_data_types.append(data_type)
                instruction_data_type_indices.append(
                    (batch_index, instruction_index, data_index)
                )
                if data_type == 4:
                    instruction_data[data_type].append(data["Payload"]["Filter"])
                else:
                    instruction_data[data_type].append(
                        vectorize_payload(data["Payload"], data_type, **factory_kwargs)
                    )
                instruction_data_indices[data_type].append(
                    (batch_index, instruction_index, data_index)
                )

    return (
        torch.tensor(instruction_types, **factory_kwargs),
        torch.tensor(instruction_indices, **factory_kwargs),
        torch.tensor(instruction_data_types, **factory_kwargs),
        torch.tensor(instruction_data_type_indices, **factory_kwargs),
        instruction_data,
        instruction_data_indices,
    )


def vectorize_payload(
    payload: dict,
    data_type: int,
    device: torch.device = None,
    dtype: torch.dtype = None,
) -> torch.Tensor:
    factory_kwargs = {"device": device, "dtype": dtype}
    match data_type:
        case 0:
            return vectorize_attack_data(payload, **factory_kwargs)
        case 1:
            return vectorize_discard_data(payload, **factory_kwargs)
        case 2:
            return vectorize_amount_data(payload, **factory_kwargs)
        case 3:
            return vectorize_return_to_deck_type_data(payload, **factory_kwargs)
        case 5:
            return vectorize_player_target_data(payload, **factory_kwargs)
        case _:
            raise ValueError(f"Unknown data type: {data_type}")

File: cpp/network/src/Nesting.cpp
#include "../include/Nesting.h"
#include "../include/TensorUtils.h"

#include <stdexcept>

namespace nesting {

namespace serialization = gamecore::serialization;

namespace {

constexpr int64_t kFilterDataType = 4;

struct TensorBuildOptions {
  std::optional<torch::Device> device{};
  std::optional<torch::Dtype> dtype{};
};

torch::TensorOptions make_options(std::optional<torch::Device> device,
                                  std::optional<torch::Dtype> dtype) {
  auto options = torch::TensorOptions();
  if (device.has_value()) {
    options = options.device(*device);
  }
  if (dtype.has_value()) {
    options = options.dtype(*dtype);
  }
  return options;
}

std::optional<int64_t> get_operator(const OperatorMap &operators,
                                    const GroupIndex &group_index) {
  auto it = operators.find(group_index);
  if (it == operators.end()) {
    return std::nullopt;
  }
  return it->second;
}

const serialization::ProtoBufFilterCondition &
require_filter_condition(const serialization::ProtoBufFilter &node) {
  if (!node.has_condition()) {
    throw std::invalid_argument("Leaf filter node is missing condition");
  }
  return node.condition();
}

std::vector<int64_t> make_condition_triplet(
    const serialization::ProtoBufFilterCondition &condition) {
  return {static_cast<int64_t>(condition.field()),
          static_cast<int64_t>(condition.operation()),
          static_cast<int64_t>(condition.value())};
}

struct FilterEntryCollector {
  explicit FilterEntryCollector(std::vector<TraverseEntry> &entries)
      : entries(entries) {}

  void append_leaf(const serialization::ProtoBufFilter &node,
                   int64_t logical_operator) {
    entries.push_back(
        TraverseEntry{make_condition_triplet(require_filter_condition(node)),
                      path, logical_operator});
  }

  void append_nested(const serialization::ProtoBufFilter &node) {
    const auto logical_operator = static_cast<int64_t>(node.logical_operator());
    const auto &operands = node.operands();
    for (int child_index = 0; child_index < operands.size(); ++child_index) {
      const auto &child = operands.Get(child_index);
      if (child.is_leaf()) {
        append_leaf(child, logical_operator);
        continue;
      }

      path.push_back(static_cast<int64_t>(child_index));
      append_nested(child);
      path.pop_back();
    }
  }

  void collect_roots(
      const std::vector<serialization::ProtoBufFilter> &nested_input) {
    for (size_t root_index = 0; root_index < nested_input.size();
         ++root_index) {
      const auto &node = nested_input[root_index];
      path.push_back(static_cast<int64_t>(root_index));
      if (node.is_leaf()) {
        append_leaf(node, 0);
      } else {
        append_nested(node);
      }
      path.pop_back();
    }
  }

  std::vector<TraverseEntry> &entries;
  GroupIndex path{};
};

GroupIndex parent_group_index(const GroupIndex &group_index) {
  GroupIndex parent = group_index;
  if (!parent.empty()) {
    parent.pop_back();
  }
  return parent;
}

struct ReduceStack {
  explicit ReduceStack(const ReduceRequest &request)
      : operators(request.operators),
        combine_function(request.combine_function) {}

  void start_next_batch_group() {
    current_groups.push_back(GroupIndex{current_batch_index});
    current_batch_index += 1;
    current_combination.push_back({});
  }

  void push_new_group(const GroupIndex &group_index,
                      const torch::Tensor &element) {
    current_combination.push_back({element});
    current_groups.push_back(group_index);
  }

  torch::Tensor combine_top_group() {
    const auto current_group = current_groups.back();
    current_groups.pop_back();
    const auto combined = combine_function(
        current_combination.back(), get_operator(operators, current_group));
    current_combination.pop_back();
    return combined;
  }

  void break_down_to(const GroupIndex &group_index) {
    while (!current_groups.empty() &&
           !is_prefix(current_groups.back(), group_index)) {
      const auto current_group = current_groups.back();
      const auto combined = combine_top_group();
      add_element(combined, parent_group_index(current_group));
    }
  }

  void add_element(const torch::Tensor &element,
                   const GroupIndex &group_index) {
    if (!current_groups.empty() && group_index == current_groups.back()) {
      current_combination.back().push_back(element);
      return;
    }

    break_down_to(group_index);

    if (!current_groups.empty() && group_index == current_groups.back()) {
      current_combination.back().push_back(element);
      return;
    }

    push_new_group(group_index, element);
  }

  std::vector<std::vector<torch::Tensor>> current_combination{};
  std::vector<GroupIndex> current_groups{};
  int64_t current_batch_index = 0;
  const OperatorMap &operators;
  const ReduceCombineFunction &combine_function;
};

std::optional<torch::Dtype>
resolve_payload_dtype(std::optional<torch::Dtype> dtype) {
  if (dtype.has_value()) {
    return dtype;
  }
  return torch::kInt64;
}

constexpr std::array<int64_t, kNumInstructionDataTypes> kPayloadWidths = {
    2, 1, 3, 2, 0, 1};

int64_t payload_width(int64_t data_type) {
  if (data_type < 0 ||
      data_type >= static_cast<int64_t>(kPayloadWidths.size())) {
    throw std::invalid_argument("Unknown payload width for data type " +
                                std::to_string(data_type));
  }
  return kPayloadWidths[static_cast<size_t>(data_type)];
}

struct PayloadTensorBuilder {
  std::array<std::vector<int64_t>, kNumInstructionDataTypes> values{};
  std::array<std::vector<int64_t>, kNumInstructionDataTypes> global_positions{};
  std::array<int64_t, kNumInstructionDataTypes> row_counts{};
};

struct MessageLocation {
  size_t batch_index = 0;
  size_t instruction_index = 0;
};

struct DataLocation {
  MessageLocation message{};
  int data_index = 0;
};

int64_t
global_data_row_index(const std::vector<int64_t> &instruction_data_types) {
  return static_cast<int64_t>(instruction_data_types.size() - 1);
}

void record_payload_position(PayloadTensorBuilder &payload_builder,
                             int64_t data_type, int64_t global_row_index) {
  payload_builder.global_positions[static_cast<size_t>(data_type)].push_back(
      global_row_index);
  payload_builder.row_counts[static_cast<size_t>(data_type)] += 1;
}

void append_payload_values(PayloadTensorBuilder &payload_builder,
                           int64_t data_type,
                           const std::vector<int64_t> &payload_values) {
  const auto width = payload_width(data_type);
  if (static_cast<int64_t>(payload_values.size()) != width) {
    throw std::invalid_argument("Payload width mismatch for instruction data");
  }

  auto &buffer = payload_builder.values[static_cast<size_t>(data_type)];
  buffer.insert(buffer.end(), payload_values.begin(), payload_values.end());
}

std::vector<int64_t>
vectorize_payload_values(const serialization::ProtoBufInstructionData &data) {
  const auto data_type = static_cast<int64_t>(data.instruction_data_type());
  switch (data_type) {
  case 0:
    if (!data.has_attack_data()) {
      throw std::invalid_argument("InstructionDataType 0 requires attack_data");
    }
    return {static_cast<int64_t>(data.attack_data().attack_target()),
            static_cast<int64_t>(data.attack_data().damage())};
  case 1:
    if (!data.has_discard_data()) {
      throw std::invalid_argument(
          "InstructionDataType 1 requires discard_data");
    }
    return {static_cast<int64_t>(data.discard_data().target_source())};
  case 2:
    if (!data.has_card_amount_data() || !data.card_amount_data().has_amount()) {
      throw std::invalid_argument(
          "InstructionDataType 2 requires card_amount_data.amount");
    }
    return {static_cast<int64_t>(data.card_amount_data().amount().min()),
            static_cast<int64_t>(data.card_amount_data().amount().max()),
            static_cast<int64_t>(data.card_amount_data().from_position())};
  case 3:
    if (!data.has_return_to_deck_type_data()) {
      throw std::invalid_argument(
          "InstructionDataType 3 requires return_to_deck_type_data");
    }
    return {
        static_cast<int64_t>(
            data.return_to_deck_type_data().return_to_deck_type()),
        static_cast<int64_t>(data.return_to_deck_type_data().from_position())};
  case 5:
    if (!data.has_player_target_data()) {
      throw std::invalid_argument(
          "InstructionDataType 5 requires player_target_data");
    }
    return {static_cast<int64_t>(data.player_target_data().player_target())};
  default:
    throw std::invalid_argument(
        "Unknown data type: " +
        std::to_string(static_cast<int64_t>(data.instruction_data_type())));
  }
}

struct DensePayloadTensorSpec {
  const std::vector<int64_t> &buffer;
  int64_t rows = 0;
  int64_t width = 0;
};

torch::Tensor build_dense_payload_tensor(const DensePayloadTensorSpec &spec,
                                         const TensorBuildOptions &options) {
  auto tensor_options =
      make_options(options.device, resolve_payload_dtype(options.dtype));
  if (spec.width == 0) {
    return torch::empty({0, 0}, tensor_options);
  }
  if (spec.rows == 0) {
    return torch::empty({0, spec.width}, tensor_options);
  }
  return torch::tensor(spec.buffer, tensor_options)
      .view({spec.rows, spec.width});
}

torch::Tensor
build_dense_payload_reorder(const PayloadTensorBuilder &payload_builder,
                            const TensorBuildOptions &options) {
  int64_t total_rows = 0;
  for (const auto row_count : payload_builder.row_counts) {
    total_rows += row_count;
  }
  std::vector<int64_t> reorder(static_cast<size_t>(total_rows), 0);
  int64_t type_base = 0;
  for (size_t type_index = 0; type_index < payload_builder.row_counts.size();
       ++type_index) {
    const auto &positions = payload_builder.global_positions[type_index];
    for (size_t local_index = 0; local_index < positions.size();
         ++local_index) {
      reorder[static_cast<size_t>(positions[local_index])] =
          type_base + static_cast<int64_t>(local_index);
    }
    type_base += payload_builder.row_counts[type_index];
  }

  return torch::tensor(reorder, make_options(options.device, torch::kLong));
}

void finalize_dense_payload_tensors(FlattenInstructionsResult &result,
                                    const PayloadTensorBuilder &payload_builder,
                                    const TensorBuildOptions &options) {
  for (size_t type_index = 0;
       type_index < result.instruction_data_tensors.size(); ++type_index) {
    result.instruction_data_tensors[type_index] = build_dense_payload_tensor(
        DensePayloadTensorSpec{payload_builder.values[type_index],
                               payload_builder.row_counts[type_index],
                               kPayloadWidths[type_index]},
        options);
  }

  result.instruction_data_reorder =
      build_dense_payload_reorder(payload_builder, options);
}

struct FilterCompileBuilder {
  std::vector<bool> node_is_leaf{};
  std::vector<int64_t> node_logical_operator{};
  std::vector<int64_t> node_depth{};
  std::vector<int64_t> child_ptr{};
  std::vector<int64_t> child_idx{};
  std::vector<int64_t> leaf_node_index{};
  std::vector<int64_t> leaf_field{};
  std::vector<int64_t> leaf_compare_op{};
  std::vector<int64_t> leaf_value{};
  std::vector<int64_t> root_node_index{};
};

void append_leaf_condition(const serialization::ProtoBufFilter &node,
                           int64_t node_id, FilterCompileBuilder &builder) {
  const auto &condition = require_filter_condition(node);
  builder.leaf_node_index.push_back(node_id);
  builder.leaf_field.push_back(static_cast<int64_t>(condition.field()));
  builder.leaf_compare_op.push_back(
      static_cast<int64_t>(condition.operation()));
  builder.leaf_value.push_back(static_cast<int64_t>(condition.value()));
}

int64_t append_compiled_filter_node(const serialization::ProtoBufFilter &node,
                                    int64_t depth,
                                    FilterCompileBuilder &builder) {
  std::vector<int64_t> child_nodes;
  child_nodes.reserve(node.operands_size());
  for (int child_index = 0; child_index < node.operands_size(); ++child_index) {
    child_nodes.push_back(append_compiled_filter_node(
        node.operands(child_index), depth + 1, builder));
  }

  const auto node_id = static_cast<int64_t>(builder.node_is_leaf.size());
  const auto child_start = static_cast<int64_t>(builder.child_idx.size());
  builder.child_idx.insert(builder.child_idx.end(), child_nodes.begin(),
                           child_nodes.end());
  builder.child_ptr.push_back(child_start);
  builder.node_is_leaf.push_back(node.is_leaf());
  builder.node_logical_operator.push_back(
      node.is_leaf() ? 0 : static_cast<int64_t>(node.logical_operator()));
  builder.node_depth.push_back(depth);

  if (node.is_leaf()) {
    append_leaf_condition(node, node_id, builder);
  }

  return node_id;
}

torch::Tensor tensor_from_bool_vector(const std::vector<bool> &values,
                                      std::optional<torch::Device> device) {
  std::vector<uint8_t> dense(values.begin(), values.end());
  return torch::tensor(dense, torch::TensorOptions().dtype(torch::kBool))
      .to(device.value_or(torch::Device(torch::kCPU)));
}

FilterBatchTensors
make_filter_batch_tensors(const FilterCompileBuilder &builder,
                          const TensorBuildOptions &options) {
  const auto index_dtype = resolve_payload_dtype(options.dtype);
  FilterBatchTensors out;
  out.node_is_leaf =
      tensor_from_bool_vector(builder.node_is_leaf, options.device);
  out.node_logical_operator = torch::tensor(
      builder.node_logical_operator, make_options(options.device, index_dtype));
  out.node_depth = torch::tensor(builder.node_depth,
                                 make_options(options.device, index_dtype));
  out.child_ptr = torch::tensor(builder.child_ptr,
                                make_options(options.device, index_dtype));
  out.child_idx = torch::tensor(builder.child_idx,
                                make_options(options.device, index_dtype));
  out.leaf_node_index = torch::tensor(
      builder.leaf_node_index, make_options(options.device, index_dtype));
  out.leaf_field = torch::tensor(builder.leaf_field,
                                 make_options(options.device, index_dtype));
  out.leaf_compare_op = torch::tensor(
      builder.leaf_compare_op, make_options(options.device, index_dtype));
  out.leaf_value = torch::tensor(builder.leaf_value,
                                 make_options(options.device, index_dtype));
  out.root_node_index = torch::tensor(
      builder.root_node_index, make_options(options.device, index_dtype));
  return out;
}

std::vector<int64_t> make_instruction_index(const MessageLocation &location) {
  return {static_cast<int64_t>(location.batch_index),
          static_cast<int64_t>(location.instruction_index)};
}

std::vector<int64_t> make_instruction_data_index(const DataLocation &location) {
  return {static_cast<int64_t>(location.message.batch_index),
          static_cast<int64_t>(location.message.instruction_index),
          static_cast<int64_t>(location.data_index)};
}

void append_vectorized_payload(
    const serialization::ProtoBufInstructionData &data, int64_t data_type,
    PayloadTensorBuilder &payload_builder) {
  append_payload_values(payload_builder, data_type,
                        vectorize_payload_values(data));
}

template <typename MessageType, typename TypeAccessor>
struct FlatMessageBuilder {
  explicit FlatMessageBuilder(const TypeAccessor &type_accessor)
      : type_accessor(type_accessor) {}

  void
  append_instruction_data(const serialization::ProtoBufInstructionData &data,
                          const DataLocation &location, int64_t parent_row) {
    const auto data_type = static_cast<int64_t>(data.instruction_data_type());
    instruction_data_types.push_back(data_type);
    instruction_data_parent_rows.push_back(parent_row);
    instruction_data_type_indices.push_back(
        make_instruction_data_index(location));

    if (data_type == kFilterDataType) {
      if (!data.has_filter_data() || !data.filter_data().has_filter()) {
        throw std::invalid_argument(
            "InstructionDataType 4 requires filter_data.filter payload");
      }
      filter_data.push_back({data.filter_data().filter()});
    } else {
      append_vectorized_payload(data, data_type, payload_builder);
    }

    // Filters (type 4) have width 0 in the payload tensor but still occupy
    // positions in the reorder mapping so their embeddings are interleaved
    // with other data-type embeddings in the original order.
    record_payload_position(payload_builder, data_type,
                            global_data_row_index(instruction_data_types));
  }

  void append_message(const MessageType &message,
                      const MessageLocation &location) {
    instruction_types.push_back(type_accessor(message));
    instruction_indices.push_back(make_instruction_index(location));
    const auto parent_row = static_cast<int64_t>(instruction_types.size() - 1);

    for (int data_index = 0; data_index < message.data_size(); ++data_index) {
      append_instruction_data(message.data(data_index),
                              DataLocation{location, data_index}, parent_row);
    }
  }

  FlattenInstructionsResult build(const TensorBuildOptions &options) {
    FlattenInstructionsResult result;
    const auto tensor_options = make_options(options.device, options.dtype);
    result.instruction_types = torch::tensor(instruction_types, tensor_options);
    result.instruction_indices = tensor_utils::tensor_from_2d_int64(
        instruction_indices, options.device, options.dtype);
    result.instruction_data_types =
        torch::tensor(instruction_data_types, tensor_options);
    result.instruction_data_parent_rows =
        torch::tensor(instruction_data_parent_rows, tensor_options);
    result.instruction_data_type_indices = tensor_utils::tensor_from_2d_int64(
        instruction_data_type_indices, options.device, options.dtype);
    finalize_dense_payload_tensors(result, payload_builder, options);
    result.filter_batch =
        compile_filter_batch(filter_data, options.device, options.dtype);
    return result;
  }

  const TypeAccessor &type_accessor;
  std::vector<int64_t> instruction_types{};
  std::vector<std::vector<int64_t>> instruction_indices{};
  std::vector<int64_t> instruction_data_types{};
  std::vector<int64_t> instruction_data_parent_rows{};
  std::vector<std::vector<int64_t>> instruction_data_type_indices{};
  std::vector<std::vector<serialization::ProtoBufFilter>> filter_data{};
  PayloadTensorBuilder payload_builder{};
};

template <typename MessageType, typename TypeAccessor>
FlattenInstructionsResult
flatten_messages(const std::vector<std::vector<MessageType>> &instruction_likes,
                 const TypeAccessor &type_accessor,
                 const TensorBuildOptions &options) {
  FlatMessageBuilder<MessageType, TypeAccessor> builder(type_accessor);

  for (size_t batch_index = 0; batch_index < instruction_likes.size();
       ++batch_index) {
    const auto &batch_instruction_likes = instruction_likes[batch_index];
    for (size_t instruction_index = 0;
         instruction_index < batch_instruction_likes.size();
         ++instruction_index) {
      builder.append_message(batch_instruction_likes[instruction_index],
                             MessageLocation{batch_index, instruction_index});
    }
  }

  return builder.build(options);
}

torch::Tensor move_tensor_if_defined(const torch::Tensor &tensor,
                                     torch::Device device) {
  return tensor.defined() ? tensor.to(device) : tensor;
}

} // namespace

bool is_prefix(const GroupIndex &prefix, const GroupIndex &test) {
  if (prefix.size() > test.size()) {
    return false;
  }
  for (size_t i = 0; i < prefix.size(); ++i) {
    if (prefix[i] != test[i]) {
      return false;
    }
  }
  return true;
}

std::vector<TraverseEntry> traverse_filter(
    const std::vector<serialization::ProtoBufFilter> &nested_input) {
  std::vector<TraverseEntry> out;
  FilterEntryCollector collector(out);
  collector.collect_roots(nested_input);
  return out;
}

FlattenResult flatten(const std::vector<TraverseEntry> &entries) {
  FlattenResult result;
  result.flattened_input.reserve(entries.size());
  result.groups.reserve(entries.size());

  for (const auto &entry : entries) {
    result.flattened_input.push_back(entry.value);
    result.groups.push_back(entry.group_index);
    result.operators[entry.group_index] = entry.op;
  }
  return result;
}

std::vector<torch::Tensor> reduce(const ReduceRequest &request) {
  ReduceStack stack(request);

  for (size_t i = 0; i < request.groups.size(); ++i) {
    if (stack.current_groups.empty()) {
      stack.start_next_batch_group();
    }

    stack.add_element(request.flattened_input[i], request.groups[i]);
  }

  stack.break_down_to(GroupIndex{});
  if (stack.current_combination.empty()) {
    return {};
  }
  return stack.current_combination.back();
}

FilterBatchTensors compile_filter_batch(
    const std::vector<std::vector<serialization::ProtoBufFilter>> &filters,
    std::optional<torch::Device> device, std::optional<torch::Dtype> dtype) {
  FilterCompileBuilder builder;
  for (const auto &forest : filters) {
    for (const auto &root : forest) {
      builder.root_node_index.push_back(
          append_compiled_filter_node(root, 0, builder));
    }
  }
  builder.child_ptr.push_back(static_cast<int64_t>(builder.child_idx.size()));
  return make_filter_batch_tensors(builder, TensorBuildOptions{device, dtype});
}

FlattenInstructionsResult flatten_instructions(
    const std::vector<std::vector<serialization::ProtoBufInstruction>>
        &instructions,
    std::optional<torch::Device> device, std::optional<torch::Dtype> dtype) {
  return flatten_messages(
      instructions,
      [](const serialization::ProtoBufInstruction &instruction) {
        return static_cast<int64_t>(instruction.instruction_type());
      },
      TensorBuildOptions{device, dtype});
}

FlattenInstructionsResult flatten_conditions(
    const std::vector<std::vector<serialization::ProtoBufCondition>>
        &conditions,
    std::optional<torch::Device> device, std::optional<torch::Dtype> dtype) {
  return flatten_messages(
      conditions,
      [](const serialization::ProtoBufCondition &condition) {
        return static_cast<int64_t>(condition.condition_type());
      },
      TensorBuildOptions{device, dtype});
}

FlattenInstructionsResult
move_flattened_result_to_device(const FlattenInstructionsResult &result,
                                torch::Device device) {
  FlattenInstructionsResult moved;
  moved.instruction_types = result.instruction_types.to(device);
  moved.instruction_indices = result.instruction_indices.to(device);
  moved.instruction_data_types = result.instruction_data_types.to(device);
  moved.instruction_data_parent_rows =
      result.instruction_data_parent_rows.to(device);
  moved.instruction_data_type_indices =
      result.instruction_data_type_indices.to(device);
  moved.instruction_data_reorder = result.instruction_data_reorder.to(device);
  moved.filter_batch = move_filter_batch_to_device(result.filter_batch, device);

  for (size_t i = 0; i < moved.instruction_data_tensors.size(); ++i) {
    moved.instruction_data_tensors[i] =
        result.instruction_data_tensors[i].to(device);
  }

  return moved;
}

FilterBatchTensors move_filter_batch_to_device(const FilterBatchTensors &batch,
                                               torch::Device device) {
  FilterBatchTensors moved;
  moved.node_is_leaf = move_tensor_if_defined(batch.node_is_leaf, device);
  moved.node_logical_operator =
      move_tensor_if_defined(batch.node_logical_operator, device);
  moved.node_depth = move_tensor_if_defined(batch.node_depth, device);
  moved.child_ptr = move_tensor_if_defined(batch.child_ptr, device);
  moved.child_idx = move_tensor_if_defined(batch.child_idx, device);
  moved.leaf_node_index = move_tensor_if_defined(batch.leaf_node_index, device);
  moved.leaf_field = move_tensor_if_defined(batch.leaf_field, device);
  moved.leaf_compare_op = move_tensor_if_defined(batch.leaf_compare_op, device);
  moved.leaf_value = move_tensor_if_defined(batch.leaf_value, device);
  moved.root_node_index = move_tensor_if_defined(batch.root_node_index, device);
  return moved;
}

} // namespace nesting
